specification/compiler/steps/validate-model.ts

```diff
@@ -0,0 +1,594 @@
+      const names = [prop.name].concat(prop.aliases ?? [])
+      for (const name of names) {
+        if (allNames.has(name)) {
+          modelError(`Duplicate property name or alias: '${name}'`)
+        }
+        allNames.add(name)
+      }
+
+      context.push(`Property '${prop.name}'`)
+      validateValueOf(prop.type, openGenerics)
+      context.pop()
+    }
+  }
+
+  // Validates that use of non-leaf types (i.e. those who are inherited or implemented) are used either
+  // in contexts that are distinct from that of their children, or are abstract parent classes providing a way
+  // way to identify actual concrete implementations in strongly typed languages.
+  function validateIsLeafType (type: model.TypeName): void {
+    if (parentTypes.has(fqn(type))) {
+      // Aggregations are disambiguated with the 'typed_keys' parameter
+      if (type.namespace === 'aggregations') return
+
+      const fqName = fqn(type)
+
+      switch (fqName) {
+        // Base type also used as property, no polymorphic usage
+        case 'internal:ErrorCause':
+        case 'x_pack.enrich:EnrichPolicy':
+        case 'x_pack.info.x_pack_usage:XPackUsage':
+        case 'x_pack.info.x_pack_usage:SecurityFeatureToggle':
+        case 'x_pack.watcher.watcher_stats:WatchRecordQueuedStats':
+        case 'x_pack.security.user.get_user:XPackUser':
+        case 'cluster.nodes_stats:MemoryStats':
+        case 'search.search:SearchResponse':
+          return
+
+        // Have a "type" attribute that identifies the variant
+        case 'mapping.types:PropertyBase':
+        case 'analysis.token_filters:TokenFilterBase':
+          return
+
+        // Single subclass with no additional properties, can probably be removed
+        case 'x_pack.watcher.input:HttpInputRequestDefinition':
+          return
+      }
+
+      modelError(`Non-leaf type cannot be used here: '${fqName}'`)
+    }
+  }
+
+  // -----------------------------------------------------------------------------------------------
+  // Value_Of validations
+
+  function validateValueOf (valueOf: model.ValueOf, openGenerics: Set<string>): void {
+    context.push(valueOf.kind)
+    switch (valueOf.kind) {
+      case 'instance_of':
+        validateTypeRef(valueOf.type, valueOf.generics, openGenerics)
+        validateIsLeafType(valueOf.type)
+        break
+
+      case 'array_of':
+        validateValueOf(valueOf.value, openGenerics)
+        break
+
+      case 'union_of':
+        for (const item of valueOf.items) {
+          validateValueOf(item, openGenerics)
+        }
+        break
+
+      case 'dictionary_of':
+        validateValueOf(valueOf.key, openGenerics)
+        validateValueOf(valueOf.value, openGenerics)
+        break
+
+      case 'named_value_of':
+        validateValueOf(valueOf.value, openGenerics)
+        break
+
+      case 'user_defined_value':
+        // Nothing to validate
+        break
+    }
```

Comment:
Let's make this future proof :)

```suggestion
      case 'user_defined_value':
        // Nothing to validate
        break
    
      default:
        throw new Error(`Unknown kind: ${valueOf.kind}`)
    }
```

Reply:
This is an interesting one... Since all `kind` are tested, this fails with `error TS2339: Property 'kind' does not exist on type 'never'.`

And omiting a `kind` does not cause a compilation error like it would do if we returned a value in the various `case` branches (like in shown in [this post](https://blog.logrocket.com/pattern-matching-and-type-safety-in-typescript-1da1231a2e34/))

This is the "correct" way to address this:

```
      default:
        // @ts-expect-error
        throw new Error(`Unknown kind: ${valueOf.kind}`)
```

Forgetting one of the `kind` values then causes the `ts-expect-error` to fail 😆

Reply:
Perfect!